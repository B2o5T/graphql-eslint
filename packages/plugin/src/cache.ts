// Based on the `eslint-plugin-import`'s cache
// https://github.com/import-js/eslint-plugin-import/blob/main/utils/ModuleCache.js
import debugFactory from 'debug';

const log = debugFactory('graphql-eslint:ModuleCache');

export class ModuleCache<K, T> {
  map = new Map<K, { lastSeen: [number, number]; result: T }>();

  set(cacheKey: K, result: T): void {
    this.map.set(cacheKey, { lastSeen: process.hrtime(), result });
    log('setting entry for', cacheKey);
  }

  get(cacheKey: K, settings = { lifetime: 10 /* seconds */ }): T | void {
    const value = this.map.get(cacheKey);
    if (!value) {
      log('cache miss for', cacheKey);
      return;
    }
    const { lastSeen, result } = value;
    // check freshness
    if (
      process.env.NODE /* don't check for ESLint CLI */ ||
      process.hrtime(lastSeen)[0] < settings.lifetime
    ) {
      return result;
    }
  }
}
