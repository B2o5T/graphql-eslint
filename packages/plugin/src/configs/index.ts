import { ConfigName } from '../types.js';
// @ts-expect-error -- complains about no default export
import operationsAllConfig from './operations-all.js';
// @ts-expect-error -- complains about no default export
import operationsRecommendedConfig from './operations-recommended.js';
// @ts-expect-error -- complains about no default export
import schemaAllConfig from './schema-all.js';
// @ts-expect-error -- complains about no default export
import schemaRecommendedConfig from './schema-recommended.js';
// @ts-expect-error -- complains about no default export
import relayConfig from './schema-relay.js';

export const configs = {
  'schema-recommended': schemaRecommendedConfig,
  'schema-all': schemaAllConfig,
  'schema-relay': relayConfig,
  'operations-recommended': operationsRecommendedConfig,
  'operations-all': operationsAllConfig,
  'flat/schema-recommended': schemaRecommendedConfig.rules,
  'flat/schema-all': {
    ...schemaRecommendedConfig.rules,
    ...schemaAllConfig.rules,
  },
  'flat/schema-relay': relayConfig.rules,
  'flat/operations-recommended': operationsRecommendedConfig.rules,
  'flat/operations-all': {
    ...operationsRecommendedConfig.rules,
    ...operationsAllConfig.rules,
  },
} satisfies Record<ConfigName | `flat/${ConfigName}`, unknown>;
