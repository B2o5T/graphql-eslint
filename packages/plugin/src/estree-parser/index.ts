export * from "./converter";
export * from "./estree-ast";
export * from "./utils";
