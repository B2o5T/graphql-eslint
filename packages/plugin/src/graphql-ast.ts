import {
  GraphQLSchema,
  GraphQLFieldMap,
  GraphQLInputFieldMap,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputType,
  GraphQLOutputType,
  GraphQLNamedType,
  GraphQLInterfaceType,
  GraphQLArgument,
  isObjectType,
  isInterfaceType,
  isUnionType,
  isInputObjectType,
  isListType,
  isNonNullType,
} from 'graphql';

export function getReachableTypes(schema: GraphQLSchema): Set<string>
export function getReachableTypes(schema?: GraphQLSchema): Set<string> | null {
  if (schema) {
    return collectReachableTypes(schema);
  }
  
  return null;
}

export function collectReachableTypes(schema: GraphQLSchema): Set<string> {
  const reachableTypes = new Set<string>();

  collectFrom(schema.getQueryType());
  collectFrom(schema.getMutationType());
  collectFrom(schema.getSubscriptionType());

  return reachableTypes;

  function collectFrom(type?: GraphQLNamedType): void {
    if (type && shouldCollect(type.name)) {
      if (isObjectType(type) || isInterfaceType(type)) {
        collectFromFieldMap(type.getFields());
        collectFromInterfaces(type.getInterfaces());
      } else if (isUnionType(type)) {
        type.getTypes().forEach(collectFrom);
      } else if (isInputObjectType(type)) {
        collectFromInputFieldMap(type.getFields());
      }
    }
  }

  function collectFromFieldMap(fieldMap: GraphQLFieldMap<any, any>): void {
    for (const fieldName in fieldMap) {
      collectFromField(fieldMap[fieldName]);
    }
  }

  function collectFromField(field: GraphQLField<any, any>): void {
    collectFromOutputType(field.type);
    field.args.forEach(collectFromArgument);
  }

  function collectFromArgument(arg: GraphQLArgument): void {
    collectFromInputType(arg.type);
  }

  function collectFromInputFieldMap(fieldMap: GraphQLInputFieldMap): void {
    for (const fieldName in fieldMap) {
      collectFromInputField(fieldMap[fieldName]);
    }
  }

  function collectFromInputField(field: GraphQLInputField): void {
    collectFromInputType(field.type);
  }

  function collectFromInterfaces(interfaces: GraphQLInterfaceType[]): void {
    if (interfaces) {
      interfaces.forEach(interfaceType => {
        collectFromFieldMap(interfaceType.getFields());
        collectFromInterfaces(interfaceType.getInterfaces());
      });
    }
  }

  function collectFromOutputType(output: GraphQLOutputType): void {
    collectFrom(schema.getType(resolveName(output)));
  }

  function collectFromInputType(input: GraphQLInputType): void {
    collectFrom(schema.getType(resolveName(input)));
  }

  function resolveName(type: GraphQLOutputType | GraphQLInputType) {
    if (isListType(type) || isNonNullType(type)) {
      return resolveName(type.ofType);
    }

    return type.name;
  }

  function shouldCollect(name: string): boolean {
    if (!reachableTypes.has(name)) {
      reachableTypes.add(name);
      return true;
    }

    return false;
  }
}
