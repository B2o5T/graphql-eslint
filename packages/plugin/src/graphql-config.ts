import { dirname } from 'path';
import debugFactory from 'debug';
import {
  GraphQLConfig,
  GraphQLExtensionDeclaration,
  loadConfigSync,
  SchemaPointer,
} from 'graphql-config';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import { ParserOptions } from './types';

const debug = debugFactory('graphql-eslint:graphql-config');
let graphQLConfig: GraphQLConfig;

export function loadGraphQLConfig(options: ParserOptions = {}): GraphQLConfig {
  // We don't want cache config on test environment
  // Otherwise schema and documents will be same for all tests
  if (process.env.NODE_ENV !== 'test' && graphQLConfig) {
    return graphQLConfig;
  }

  const onDiskConfig = options.skipGraphQLConfig
    ? null
    : loadConfigSync({
        // load config relative to the file being linted
        rootDir: options.filePath ? dirname(options.filePath) : undefined,
        throwOnEmpty: false,
        throwOnMissing: false,
        extensions: [addCodeFileLoaderExtension],
      });

  debug('options.skipGraphQLConfig: %o', options.skipGraphQLConfig);
  if (onDiskConfig) {
    debug('Graphql-config path %o', onDiskConfig.filepath);
  }

  const configOptions = options.projects
    ? { projects: options.projects }
    : {
        schema: (options.schema || '') as SchemaPointer, // if `schema` is `undefined` will throw error `Project 'default' not found`
        documents: options.documents || options.operations,
        extensions: options.extensions,
        include: options.include,
        exclude: options.exclude,
      };

  graphQLConfig =
    onDiskConfig ||
    new GraphQLConfig(
      {
        config: configOptions,
        filepath: 'virtual-config',
      },
      [addCodeFileLoaderExtension]
    );

  return graphQLConfig;
}

const addCodeFileLoaderExtension: GraphQLExtensionDeclaration = api => {
  api.loaders.schema.register(new CodeFileLoader());
  api.loaders.documents.register(new CodeFileLoader());
  return { name: 'graphql-eslint-loaders' };
};
