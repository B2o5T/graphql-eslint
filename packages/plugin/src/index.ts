import { configs } from './configs/index.js';
import { parseForESLint, parser } from './parser.js';
import { processor } from './processor.js';
import { rules } from './rules/index.js';

export * from './types.js';
export type { IGraphQLConfig } from 'graphql-config';
export type { GraphQLTagPluckOptions } from '@graphql-tools/graphql-tag-pluck';

export { requireGraphQLSchemaFromContext, requireSiblingsOperations } from './utils.js';

export const processors = { graphql: processor };

export { rules, configs, parser, parseForESLint };

// eslint-disable-next-line import/no-default-export -- It's common practice for ESLint plugins that supports Flat config to use the default export
export default {
  parser,
  processor,
  rules,
  configs,
};
