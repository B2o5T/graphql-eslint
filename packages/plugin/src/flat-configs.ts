import { parseForESLint } from './parser';
import { rules } from './rules';
import { configs } from './configs';

const baseOptions = {
  parser: { parseForESLint },
  plugins: {
    '@graphql-eslint': { rules },
  },
};

export const flatConfigs = {
  'operations-all': {
    ...baseOptions,
    rules: {
      ...configs['operations-recommended'].rules,
      ...configs['operations-all'].rules,
    },
  },
  'operations-recommended': {
    ...baseOptions,
    rules: configs['operations-recommended'].rules,
  },
  relay: {
    ...baseOptions,
    rules: configs.relay.rules,
  },
  'schema-all': {
    ...baseOptions,
    rules: {
      ...configs['schema-recommended'].rules,
      ...configs['schema-all'].rules,
    },
  },
  'schema-recommended': {
    ...baseOptions,
    rules: configs['schema-recommended'].rules,
  },
};
