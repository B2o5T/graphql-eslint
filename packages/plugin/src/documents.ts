import { resolve } from 'path';
import {
  FragmentDefinitionNode,
  Kind,
  OperationDefinitionNode,
  SelectionSetNode,
  visit,
  OperationTypeNode,
} from 'graphql';
import { Source } from '@graphql-tools/utils';
import { GraphQLProjectConfig } from 'graphql-config';
import debugFactory from 'debug';
import fg from 'fast-glob';
import { logger } from './utils.js';
import { Pointer } from './types.js';
import { ModuleCache } from './cache.js';

export type FragmentSource = { filePath: string; document: FragmentDefinitionNode };
export type OperationSource = { filePath: string; document: OperationDefinitionNode };

const debug = debugFactory('graphql-eslint:operations');

export type SiblingOperations = {
  available: boolean;
  getFragment(fragmentName: string): FragmentSource[];
  getFragments(): FragmentSource[];
  getFragmentByType(typeName: string): FragmentSource[];
  getFragmentsInUse(
    baseOperation: OperationDefinitionNode | FragmentDefinitionNode | SelectionSetNode,
    recursive?: boolean,
  ): FragmentDefinitionNode[];
  getOperation(operationName: string): OperationSource[];
  getOperations(): OperationSource[];
  getOperationByType(operationType: OperationTypeNode): OperationSource[];
};

const handleVirtualPath = (documents: Source[]): Source[] => {
  const filepathMap: Record<string, number> = Object.create(null);

  return documents.map(source => {
    const location = source.location!;
    if (['.gql', '.graphql'].some(extension => location.endsWith(extension))) {
      return source;
    }
    filepathMap[location] ??= -1;
    const index = (filepathMap[location] += 1);
    return {
      ...source,
      location: resolve(location, `${index}_document.graphql`),
    };
  });
};

const operationsCache = new ModuleCache<Source[]>();
const siblingOperationsCache = new Map<Source[], SiblingOperations>();

const getSiblings = (project: GraphQLProjectConfig): Source[] => {
  const documentsKey = project.documents;
  if (!documentsKey) {
    return [];
  }

  let siblings = operationsCache.get(documentsKey);

  if (!siblings) {
    debug('Loading operations from %o', project.documents);
    const documents = project.loadDocumentsSync(project.documents, {
      skipGraphQLImport: true,
      pluckConfig: project.extensions.pluckConfig,
    });
    if (debug.enabled) {
      debug('Loaded %d operations', documents.length);
      const operationsPaths = fg.sync(project.documents as Pointer, { absolute: true });
      debug('Operations pointers %O', operationsPaths);
    }
    siblings = handleVirtualPath(documents);
    operationsCache.set(documentsKey, siblings);
  }

  return siblings;
};

export function getDocuments(project: GraphQLProjectConfig): SiblingOperations {
  const siblings = getSiblings(project);

  if (siblings.length === 0) {
    let printed = false;

    const noopWarn = () => {
      if (!printed) {
        logger.warn(
          'getSiblingOperations was called without any operations. Make sure to set "parserOptions.operations" to make this feature available!',
        );
        printed = true;
      }
      return [];
    };

    return {
      available: false,
      getFragment: noopWarn,
      getFragments: noopWarn,
      getFragmentByType: noopWarn,
      getFragmentsInUse: noopWarn,
      getOperation: noopWarn,
      getOperations: noopWarn,
      getOperationByType: noopWarn,
    };
  }

  // Since the siblings array is cached, we can use it as cache key.
  // We should get the same array reference each time we get
  // to this point for the same graphql project
  const value = siblingOperationsCache.get(siblings);

  if (value) {
    return value;
  }

  let fragmentsCache: FragmentSource[] | null = null;

  const getFragments = (): FragmentSource[] => {
    if (fragmentsCache === null) {
      const result: FragmentSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.FRAGMENT_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      fragmentsCache = result;
    }
    return fragmentsCache;
  };

  let cachedOperations: OperationSource[] | null = null;

  const getOperations = (): OperationSource[] => {
    if (cachedOperations === null) {
      const result: OperationSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.OPERATION_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      cachedOperations = result;
    }
    return cachedOperations;
  };

  const getFragment = (name: string) => getFragments().filter(f => f.document.name?.value === name);

  const collectFragments = (
    selectable: SelectionSetNode | OperationDefinitionNode | FragmentDefinitionNode,
    recursive: boolean,
    collected = new Map<string, FragmentDefinitionNode>(),
  ) => {
    visit(selectable, {
      FragmentSpread(spread) {
        const fragmentName = spread.name.value;
        const [fragment] = getFragment(fragmentName);

        if (!fragment) {
          logger.warn(
            `Unable to locate fragment named "${fragmentName}", please make sure it's loaded using "parserOptions.operations"`,
          );
          return;
        }
        if (!collected.has(fragmentName)) {
          collected.set(fragmentName, fragment.document);
          if (recursive) {
            collectFragments(fragment.document, recursive, collected);
          }
        }
      },
    });
    return collected;
  };

  const siblingOperations: SiblingOperations = {
    available: true,
    getFragment,
    getFragments,
    getFragmentByType: typeName =>
      getFragments().filter(f => f.document.typeCondition?.name?.value === typeName),
    getFragmentsInUse: (selectable, recursive = true) =>
      Array.from(collectFragments(selectable, recursive).values()),
    getOperation: name => getOperations().filter(o => o.document.name?.value === name),
    getOperations,
    getOperationByType: type => getOperations().filter(o => o.document.operation === type),
  };

  siblingOperationsCache.set(siblings, siblingOperations);
  return siblingOperations;
}
