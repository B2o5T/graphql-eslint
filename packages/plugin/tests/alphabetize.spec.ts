import { GraphQLRuleTester } from '../src';
import rule, { AlphabetizeConfig } from '../src/rules/alphabetize';

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests<[AlphabetizeConfig]>('alphabetize', rule, {
  valid: [
    {
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        type User {
          age: Int
          firstName: String!
          lastName: String!
          password: String
        }
      `,
    },
    {
      options: [{ fields: ['InputObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        input UserInput {
          age: Int
          firstName: String!
          lastName: String!
          password: String
          zip: String
        }
      `,
    },
    {
      options: [{ values: ['EnumTypeDefinition'] }],
      code: /* GraphQL */ `
        enum Role {
          ADMIN
          GOD
          SUPER_ADMIN
          USER
        }
      `,
    },
    {
      name: 'should not report error if selection is duplicated',
      options: [{ selections: ['OperationDefinition'] }],
      code: /* GraphQL */ `
        query {
          test {
            a
            a
            b
          }
        }
      `,
    },
  ],
  invalid: [
    {
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        type User {
          password: String
          firstName: String!
          age: Int
          lastName: String!
        }
      `,
      errors: [
        { message: '"firstName" should be before "password"' },
        { message: '"age" should be before "firstName"' },
      ],
    },
    {
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        extend type User {
          age: Int
          firstName: String!
          password: String
          lastName: String!
        }
      `,
      errors: [{ message: '"lastName" should be before "password"' }],
    },
    {
      options: [{ fields: ['InterfaceTypeDefinition'] }],
      code: /* GraphQL */ `
        interface Test {
          cc: Int
          bb: Int
          aa: Int
        }
      `,
      errors: [{ message: '"bb" should be before "cc"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      options: [{ fields: ['InputObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        input UserInput {
          password: String
          firstName: String!
          age: Int
          lastName: String!
        }
      `,
      errors: [
        { message: '"firstName" should be before "password"' },
        { message: '"age" should be before "firstName"' },
      ],
    },
    {
      options: [{ fields: ['InputObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        extend input UserInput {
          age: Int
          firstName: String!
          password: String
          lastName: String!
        }
      `,
      errors: [{ message: '"lastName" should be before "password"' }],
    },
    {
      options: [{ values: ['EnumTypeDefinition'] }],
      code: /* GraphQL */ `
        enum Role {
          SUPER_ADMIN
          ADMIN
          USER
          GOD
        }
      `,
      errors: [{ message: '"ADMIN" should be before "SUPER_ADMIN"' }, { message: '"GOD" should be before "USER"' }],
    },
    {
      options: [{ values: ['EnumTypeDefinition'] }],
      code: /* GraphQL */ `
        extend enum Role {
          ADMIN
          SUPER_ADMIN
          GOD
          USER
        }
      `,
      errors: [{ message: '"GOD" should be before "SUPER_ADMIN"' }],
    },
    {
      options: [{ arguments: ['DirectiveDefinition'] }],
      code: /* GraphQL */ `
        directive @test(cc: Int, bb: Int, aa: Int) on FIELD_DEFINITION
      `,
      errors: [{ message: '"bb" should be before "cc"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      options: [{ arguments: ['FieldDefinition'] }],
      code: /* GraphQL */ `
        type Query {
          test(cc: Int, bb: Int, aa: Int): Int
        }
      `,
      errors: [{ message: '"bb" should be before "cc"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      options: [{ selections: ['FragmentDefinition'] }],
      code: /* GraphQL */ `
        fragment TestFields on Test {
          cc
          bb
          aa
        }
      `,
      errors: [{ message: '"bb" should be before "cc"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      options: [{ selections: ['OperationDefinition'] }],
      code: /* GraphQL */ `
        query {
          test {
            cc
            bb
            aa
            ... on Test {
              ccc
              bbb
              aaa
            }
          }
        }
      `,
      errors: [
        { message: '"bb" should be before "cc"' },
        { message: '"aa" should be before "bb"' },
        { message: '"bbb" should be before "ccc"' },
        { message: '"aaa" should be before "bbb"' },
      ],
    },
    {
      options: [{ variables: ['OperationDefinition'], arguments: ['Field'] }],
      code: /* GraphQL */ `
        mutation ($cc: Int, $bb: Int, $aa: Int) {
          test(ccc: $cc, bbb: $bb, aaa: $aa) {
            something
          }
        }
      `,
      errors: [
        { message: '"$bb" should be before "$cc"' },
        { message: '"$aa" should be before "$bb"' },
        { message: '"bbb" should be before "ccc"' },
        { message: '"aaa" should be before "bbb"' },
      ],
    },
    {
      name: 'should not autofix for fields with comment between',
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        type Test {
          cc: Int
          # between 'cc' and 'c'
          c: Float
          bb: String
          aa: Int
        }
      `,
      errors: [{ message: '"c" should be before "cc"' },{ message: '"bb" should be before "c"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      name: 'should autofix if after comment is not on same line',
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        type Test {
          cc: Int
          c: Float
          # ok for 'cc' and 'c' but not for 'bb' and 'aa'
          bb: String
          aa: Int
        }
      `,
      errors: [{ message: '"c" should be before "cc"' },{ message: '"bb" should be before "c"' }, { message: '"aa" should be before "bb"' }],
    },
    {
      name: 'should autofix if before comment is not on same line',
      options: [{ fields: ['ObjectTypeDefinition'] }],
      code: /* GraphQL */ `
        type Test {
          cc: Int
          c: Float # ok for 'bb' and 'aa' but not for 'cc' and 'c'
          bb: String
          aa: Int
        }
      `,
      errors: [{ message: '"c" should be before "cc"' },{ message: '"bb" should be before "c"' }, { message: '"aa" should be before "bb"' }],
    },
  ],
});
