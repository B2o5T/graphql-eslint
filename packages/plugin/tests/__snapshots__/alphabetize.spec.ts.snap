// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Error 1/2

  1 |
  2 |         type User {
  3 |           password: String
> 4 |           firstName: String!
    |           ^^^^^^^^^ "firstName" should be before "password"
  5 |           age: Int
  6 |           lastName: String!
  7 |         }
  8 |       

Error 2/2

  1 |
  2 |         type User {
  3 |           password: String
  4 |           firstName: String!
> 5 |           age: Int
    |           ^^^ "age" should be before "firstName"
  6 |           lastName: String!
  7 |         }
  8 |       

Autofix output

type User {
  age: Int
  firstName: String!
  lastName: String!
  password: String
}
`;

exports[` 2`] = `
Error 1/1

  1 |
  2 |         extend type User {
  3 |           age: Int
  4 |           firstName: String!
  5 |           password: String
> 6 |           lastName: String!
    |           ^^^^^^^^ "lastName" should be before "password"
  7 |         }
  8 |       

Autofix output

extend type User {
  age: Int
  firstName: String!
  lastName: String!
  password: String
}
`;

exports[` 3`] = `
Error 1/2

  1 |
  2 |         interface Test {
  3 |           cc: Int
> 4 |           bb: Int
    |           ^^ "bb" should be before "cc"
  5 |           aa: Int
  6 |         }
  7 |       

Error 2/2

  1 |
  2 |         interface Test {
  3 |           cc: Int
  4 |           bb: Int
> 5 |           aa: Int
    |           ^^ "aa" should be before "bb"
  6 |         }
  7 |       

Autofix output

interface Test {
  aa: Int
  bb: Int
  cc: Int
}
`;

exports[` 4`] = `
Error 1/2

  1 |
  2 |         input UserInput {
  3 |           password: String
> 4 |           firstName: String!
    |           ^^^^^^^^^ "firstName" should be before "password"
  5 |           age: Int
  6 |           lastName: String!
  7 |         }
  8 |       

Error 2/2

  1 |
  2 |         input UserInput {
  3 |           password: String
  4 |           firstName: String!
> 5 |           age: Int
    |           ^^^ "age" should be before "firstName"
  6 |           lastName: String!
  7 |         }
  8 |       

Autofix output

input UserInput {
  age: Int
  firstName: String!
  lastName: String!
  password: String
}
`;

exports[` 5`] = `
Error 1/1

  1 |
  2 |         extend input UserInput {
  3 |           age: Int
  4 |           firstName: String!
  5 |           password: String
> 6 |           lastName: String!
    |           ^^^^^^^^ "lastName" should be before "password"
  7 |         }
  8 |       

Autofix output

extend input UserInput {
  age: Int
  firstName: String!
  lastName: String!
  password: String
}
`;

exports[` 6`] = `
Error 1/2

  1 |
  2 |         enum Role {
  3 |           SUPER_ADMIN
> 4 |           ADMIN
    |           ^^^^^ "ADMIN" should be before "SUPER_ADMIN"
  5 |           USER
  6 |           GOD
  7 |         }
  8 |       

Error 2/2

  1 |
  2 |         enum Role {
  3 |           SUPER_ADMIN
  4 |           ADMIN
  5 |           USER
> 6 |           GOD
    |           ^^^ "GOD" should be before "USER"
  7 |         }
  8 |       

Autofix output

enum Role {
  ADMIN
  GOD
  SUPER_ADMIN
  USER
}
`;

exports[` 7`] = `
Error 1/1

  1 |
  2 |         extend enum Role {
  3 |           ADMIN
  4 |           SUPER_ADMIN
> 5 |           GOD
    |           ^^^ "GOD" should be before "SUPER_ADMIN"
  6 |           USER
  7 |         }
  8 |       

Autofix output

extend enum Role {
  ADMIN
  GOD
  SUPER_ADMIN
  USER
}
`;

exports[` 8`] = `
Error 1/2

  1 |
> 2 |         directive @test(cc: Int, bb: Int, aa: Int) on FIELD_DEFINITION
    |                                  ^^ "bb" should be before "cc"
  3 |       

Error 2/2

  1 |
> 2 |         directive @test(cc: Int, bb: Int, aa: Int) on FIELD_DEFINITION
    |                                           ^^ "aa" should be before "bb"
  3 |       

Autofix output

directive @test(aa: Int, bb: Int, cc: Int) on FIELD_DEFINITION
`;

exports[` 9`] = `
Error 1/2

  1 |
  2 |         type Query {
> 3 |           test(cc: Int, bb: Int, aa: Int): Int
    |                         ^^ "bb" should be before "cc"
  4 |         }
  5 |       

Error 2/2

  1 |
  2 |         type Query {
> 3 |           test(cc: Int, bb: Int, aa: Int): Int
    |                                  ^^ "aa" should be before "bb"
  4 |         }
  5 |       

Autofix output

type Query {
  test(aa: Int, bb: Int, cc: Int): Int
}
`;

exports[` 10`] = `
Error 1/2

  1 |
  2 |         fragment TestFields on Test {
  3 |           cc
> 4 |           bb
    |           ^^ "bb" should be before "cc"
  5 |           aa
  6 |         }
  7 |       

Error 2/2

  1 |
  2 |         fragment TestFields on Test {
  3 |           cc
  4 |           bb
> 5 |           aa
    |           ^^ "aa" should be before "bb"
  6 |         }
  7 |       

Autofix output

fragment TestFields on Test {
  aa
  bb
  cc
}
`;

exports[` 11`] = `
Error 1/4

   1 |
   2 |         query {
   3 |           test {
   4 |             cc
>  5 |             bb
     |             ^^ "bb" should be before "cc"
   6 |             aa
   7 |             ... on Test {
   8 |               ccc
   9 |               bbb
  10 |               aaa
  11 |             }
  12 |           }
  13 |         }
  14 |       

Error 2/4

   1 |
   2 |         query {
   3 |           test {
   4 |             cc
   5 |             bb
>  6 |             aa
     |             ^^ "aa" should be before "bb"
   7 |             ... on Test {
   8 |               ccc
   9 |               bbb
  10 |               aaa
  11 |             }
  12 |           }
  13 |         }
  14 |       

Error 3/4

   1 |
   2 |         query {
   3 |           test {
   4 |             cc
   5 |             bb
   6 |             aa
   7 |             ... on Test {
   8 |               ccc
>  9 |               bbb
     |               ^^^ "bbb" should be before "ccc"
  10 |               aaa
  11 |             }
  12 |           }
  13 |         }
  14 |       

Error 4/4

   1 |
   2 |         query {
   3 |           test {
   4 |             cc
   5 |             bb
   6 |             aa
   7 |             ... on Test {
   8 |               ccc
   9 |               bbb
> 10 |               aaa
     |               ^^^ "aaa" should be before "bbb"
  11 |             }
  12 |           }
  13 |         }
  14 |       

Autofix output

query {
  test {
    aa
    bb
    cc
    ... on Test {
      aaa
      bbb
      ccc
    }
  }
}
`;

exports[` 12`] = `
Error 1/4

  1 |
> 2 |         mutation ($cc: Int, $bb: Int, $aa: Int) {
    |                             ^^^ "$bb" should be before "$cc"
  3 |           test(ccc: $cc, bbb: $bb, aaa: $aa) {
  4 |             something
  5 |           }
  6 |         }
  7 |       

Error 2/4

  1 |
> 2 |         mutation ($cc: Int, $bb: Int, $aa: Int) {
    |                                       ^^^ "$aa" should be before "$bb"
  3 |           test(ccc: $cc, bbb: $bb, aaa: $aa) {
  4 |             something
  5 |           }
  6 |         }
  7 |       

Error 3/4

  1 |
  2 |         mutation ($cc: Int, $bb: Int, $aa: Int) {
> 3 |           test(ccc: $cc, bbb: $bb, aaa: $aa) {
    |                          ^^^ "bbb" should be before "ccc"
  4 |             something
  5 |           }
  6 |         }
  7 |       

Error 4/4

  1 |
  2 |         mutation ($cc: Int, $bb: Int, $aa: Int) {
> 3 |           test(ccc: $cc, bbb: $bb, aaa: $aa) {
    |                                    ^^^ "aaa" should be before "bbb"
  4 |             something
  5 |           }
  6 |         }
  7 |       

Autofix output

mutation ($aa: Int, $bb: Int, $cc: Int) {
  test(aaa: $aa, bbb: $bb, ccc: $cc) {
    something
  }
}
`;

exports[` 13`] = `
Error 1/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           # between 'cc' and 'c'
> 5 |           c: Float
    |           ^ "c" should be before "cc"
  6 |           bb: String
  7 |           aa: Int
  8 |         }
  9 |       

Error 2/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           # between 'cc' and 'c'
  5 |           c: Float
> 6 |           bb: String
    |           ^^ "bb" should be before "c"
  7 |           aa: Int
  8 |         }
  9 |       

Error 3/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           # between 'cc' and 'c'
  5 |           c: Float
  6 |           bb: String
> 7 |           aa: Int
    |           ^^ "aa" should be before "bb"
  8 |         }
  9 |       

Autofix output

type Test {
  cc: Int
  # between 'cc' and 'c'
  c: Float
  aa: Int
  bb: String
}
`;

exports[` 14`] = `
Error 1/3

  1 |
  2 |         type Test {
  3 |           cc: Int
> 4 |           c: Float
    |           ^ "c" should be before "cc"
  5 |           # ok for 'cc' and 'c' but not for 'bb' and 'aa'
  6 |           bb: String
  7 |           aa: Int
  8 |         }
  9 |       

Error 2/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           c: Float
  5 |           # ok for 'cc' and 'c' but not for 'bb' and 'aa'
> 6 |           bb: String
    |           ^^ "bb" should be before "c"
  7 |           aa: Int
  8 |         }
  9 |       

Error 3/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           c: Float
  5 |           # ok for 'cc' and 'c' but not for 'bb' and 'aa'
  6 |           bb: String
> 7 |           aa: Int
    |           ^^ "aa" should be before "bb"
  8 |         }
  9 |       

Autofix output

type Test {
  c: Float
  cc: Int
  # ok for 'cc' and 'c' but not for 'bb' and 'aa'
  bb: String
  aa: Int
}
`;

exports[` 15`] = `
Error 1/3

  1 |
  2 |         type Test {
  3 |           cc: Int
> 4 |           c: Float # ok for 'bb' and 'aa' but not for 'cc' and 'c'
    |           ^ "c" should be before "cc"
  5 |           bb: String
  6 |           aa: Int
  7 |         }
  8 |       

Error 2/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           c: Float # ok for 'bb' and 'aa' but not for 'cc' and 'c'
> 5 |           bb: String
    |           ^^ "bb" should be before "c"
  6 |           aa: Int
  7 |         }
  8 |       

Error 3/3

  1 |
  2 |         type Test {
  3 |           cc: Int
  4 |           c: Float # ok for 'bb' and 'aa' but not for 'cc' and 'c'
  5 |           bb: String
> 6 |           aa: Int
    |           ^^ "aa" should be before "bb"
  7 |         }
  8 |       

Autofix output

type Test {
  cc: Int
  c: Float # ok for 'bb' and 'aa' but not for 'cc' and 'c'
  aa: Int
  bb: String
}
`;
